util/unifyfs-stage/src/unifyfs-stage-transfer.c

```diff
@@ -0,0 +1,403 @@
+        return EINVAL;
+    }
+
+    fp = fopen(ctx->manifest_file, "r");
+    if (!fp) {
+        fprintf(stderr, "failed to open file %s: %s\n",
+		ctx->manifest_file, strerror(errno));
+        ret = errno;
+        goto out;
+    }
+
+    while (NULL != fgets(linebuf, LINE_MAX-1, fp)) {
+        if (strlen(linebuf) < 5) {
+	    // the manifest file perhaps ends with a couple of characters
+	    // and/or a newline not meant to be a transfer spec.
+            if (linebuf[0] == '\n') {
+	        goto out;
+            } else{
+	        fprintf(stderr, "Short (bad) manifest file line: >%s<\n",
+			linebuf);
+		ret = -EINVAL;
+                goto out;
+            }
+	}
+	ret = unifyfs_parse_manifest_line(linebuf, &src, &dst);
+	if (ret < 0) {
+	    fprintf(stderr, "failed to parse %s (%s)\n",
+		    linebuf, strerror(ret));
+            goto out;
+	}
+        if (ctx->mode == UNIFYFS_STAGE_SERIAL) {
+            if (count % total_ranks == rank) {
+                if (verbose) {
+                    fprintf(stdout, "[%d] serial transfer: src=%s, dst=%s\n",
+                                    rank, src, dst);
+                }
+
+                ret = unifyfs_transfer_file_serial(src, dst);
+                if (ret) {
+                    goto out;
+                }
+
+                if (ret < 0) {
+                    fprintf(stderr, "stat on %s failed (err=%d, %s)\n",
+                                    dst, errno, strerror(errno));
+                    ret = errno;
+                    goto out;
+                }
+
+                if (ctx->checksum) {
+                    ret = verify_checksum(src, dst);
+                    if (ret) {
+                        fprintf(stderr, "checksums for >%s< and >%s< differ!\n",
+                                src, dst);
+                        goto out;
+                    }
+                }
+            }
+        } else {
+            if (0 == rank) {
+                int fd = -1;
+
+                if (verbose) {
+                    fprintf(stdout, "[%d] parallel transfer: src=%s, dst=%s\n",
+                                    rank, src, dst);
+                }
+
+                /* FIXME: Since we cannot use the mpi barrier inside the
+                 * unifyfs_transfer_file_parallel(), we need to create the file
+                 * here before others start to access. It would be better if we
+                 * can skip this step.
+                 */
+                fd = open(dst, O_WRONLY|O_CREAT|O_TRUNC, 0600);
+                if (fd < 0) {
+                    fprintf(stderr, "[%d] failed to create the file %s\n",
+                                    rank, dst);
+                    goto out;
+                }
+
+                close(fd);
+            }
+
+            MPI_Barrier(MPI_COMM_WORLD);
+
+            ret = unifyfs_transfer_file_parallel(src, dst);
+            if (ret) {
+                goto out;
+            }
+
+            MPI_Barrier(MPI_COMM_WORLD);
```

Comment:
What is the reason for these MPI_Barriers()?

Reply:
The first barrier is to ensure the shared file is created by the rank 0. Second one is to wait until all processes finish transfer of the file, before checking the result (checksum).

Reply:
Got it, thanks for the info.